kv/memberlist/memberlist_client.go

```diff
@@ -642,52 +640,15 @@ func (m *KV) get(key string, codec codec.Codec) (out interface{}, version uint,
 		_, _ = v.value.RemoveTombstones(time.Time{})
 	}
 
-	return v.value, v.version, nil
+	return v.value, v.version
 }
 
 // WatchKey watches for value changes for given key. When value changes, 'f' function is called with the
 // latest value. Notifications that arrive while 'f' is running are coalesced into one subsequent 'f' call.
 //
 // Watching ends when 'f' returns false, context is done, or this client is shut down.
-func (m *KV) WatchKey(ctx context.Context, key string, codec codec.Codec, f func(interface{}) bool) {
-	// keep one extra notification, to avoid missing notification if we're busy running the function
-	w := make(chan string, 1)
-
-	// register watcher
-	m.watchersMu.Lock()
-	m.watchers[key] = append(m.watchers[key], w)
-	m.watchersMu.Unlock()
-
-	defer func() {
-		// unregister watcher on exit
-		m.watchersMu.Lock()
-		defer m.watchersMu.Unlock()
-
-		removeWatcherChannel(key, w, m.watchers)
-	}()
-
-	for {
-		select {
-		case <-w:
-			// value changed
-			val, _, err := m.get(key, codec)
-			if err != nil {
-				level.Warn(m.logger).Log("msg", "failed to decode value while watching for changes", "key", key, "err", err)
-				continue
-			}
-
-			if !f(val) {
-				return
-			}
-
-		case <-m.shutdown:
-			// stop watching on shutdown
-			return
-
-		case <-ctx.Done():
-			return
-		}
-	}
+func (m *KV) WatchKey(ctx context.Context, key string, f func(interface{}) bool) {
+	m.watcher.WatchKey(ctx, key, f)
```

Comment:
The change here as well as the one in `WatchPrefix` have the side effect that the new value passed to different watchers will be same. Previously each watcher will get a different deep copy of the value. 

This doesn't have implications within dskit because only `Ring` uses either of those methods and it uses only one at a time 

https://github.com/grafana/dskit/blob/4bbe2b279599a0af2507518c5e2b3faf7adb894a/ring/ring.go#L304

The deeply copied values aren't a document behaviour either, so clients shouldn't be relying on it.